core/src/smartcontracts/isi/account.rs

```diff
@@ -570,31 +558,31 @@ pub mod query {
         }
     }
 
-    impl ValidQuery for FindAccountsWithAsset {
+    impl ValidIterableQuery for FindAccountsWithAsset {
         #[metrics(+"find_accounts_with_asset")]
         fn execute<'state>(
-            &self,
+            self,
+            filter: CompoundPredicate<AccountPredicateBox>,
             state_ro: &'state impl StateReadOnly,
-        ) -> Result<Box<dyn Iterator<Item = Account> + 'state>, Error> {
+        ) -> std::result::Result<impl Iterator<Item = Account> + 'state, Error> {
```

Comment:
`+ 'state` is too restrictive (it misleadingly means that the Iterator has to live _at least_ as long as the state, not _at most_ as long). It might be worth relaxing this bound using [the captures trick](https://rust-lang.github.io/rfcs/3498-lifetime-capture-rules-2024.html#the-captures-trick), which is strictly more correct:

```rust
trait Capture<U> {}
impl<T: ?Sized, U> Capture<U> for T {}

impl Iterator<Item = Account> + 'state -> impl Iterator<Item = Account> + Capture<&'state ()>
```

Reply:
Huh, this is an unexpected sharp edge, thanks for pointing this out.

I am not sure if we should use the trick though: AFAIU, this leads to problems only if we use multiple lifetime parameters or have use another type with a different captured lifetime. We don't have any compile-time errors, so I am a bit hesitant to complicate the signature..

Reply:
Agree, no need if this hasn't caused issues so far